### PR TITLE
Converts the seed job xml to a groovy DSL

### PR DESCRIPTION
**Pull request by pomegranited**
Replaces templates/seed_job_template.xml with templates/seedJob.groovy, which will be easier to maintain.

However, see notes below for the drawbacks.

**JIRA tickets**: Implements [OLIVE-34](https://openedx.atlassian.net/browse/OLIVE-34), [OC-1528](https://tasks.opencraft.com/browse/OC-1528)

**Dependencies**: None

**Testing instructions**:

_Note: If you've already created a Jenkins Analytics vagrant instance to test PR #2959 , then you can just checkout this branch and run `vagrant provision`._

To test via vagrant:
1. Install edx/configuration requirements in a virtualenv:
   
   ```
   virtualenv venv
   . venv/bin/activate
   git clone https://git@github.com/open-craft/configuration.git
   cd configuration
   git checkout -b jill/olive-34-seed-job-groovy origin/jill/olive-34-seed-job-groovy
   make requirements
   ```
2. Create a YAML file containing the jenkins analytics configuration.
   - See [jenkins_analytics README](https://github.com/open-craft/configuration/blob/jill/olive-34-seed-job-groovy/playbooks/roles/jenkins_analytics/README.md) for details.
   - Also see [example ansible-jenkins-credentials.yml.zip](https://github.com/edx/configuration/files/219328/seed.job.ansible-jenkins-credentials.yml.zip)
3. Verify that you have access to https://github.com/edx-ops/edx-jenkins-job-dsl/
4. Build the vagrant instance, using the YAML file created above:
   
   ```
   cd vagrant/base/analytics_jenkins
   export VAGRANT_JENKINS_LOCAL_VARS_FILE=/path/to/your.yml
   vagrant up
   ```
5. Should run successfully, building the AnalyticsSeedJob successfully as the last step.

**Author notes and concerns**:

I had issues running the seedJob.groovy file, and I'm not happy with the solution I ended up with.  Would appreciate any ideas, if you have time.
- Tried running directly via Jenkins CLI groovy and the [jenkins /script console](https://wiki.jenkins-ci.org/display/JENKINS/Jenkins+Script+Console).  I was able to import javaposse.jobdsl classes, but unable to create any jobs?  Got errors like `groovy.lang.MissingMethodException: No signature of method: RemoteClass.FreeStyleJob()`.  
- Tried wrapping it in a call to javaposse.jobdsl.Run.main and running via Jenkins
  CLI.  But this class tries to write to './AnalyticsSeedJob.xml', and errors out, even when running as root!  Baffled.

So instead, tasks/main.yml clones the job-dsl-plugin repo, and generates the
job XML using the instructions in the [job-dsl wiki](https://github.com/jenkinsci/job-dsl-plugin/wiki/User-Power-Moves).  The  build takes 3-7 min the first time, and <30 sec each subsequent run.

**Reviewers**
- [ ] @jbzdak 
- [ ] @mulby 
- [ ] @jibsheet

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @pomegranited! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=2963&repo=edx%2Fconfiguration

File: playbooks/roles/jenkins_analytics/defaults/main.yml [hunk 1]
```diff
@@ -89,8 +93,32 @@ jenkins_seed_job:
       dest: "{{ ANALYTICS_SCHEDULE_SECURE_REPO_DEST | default('') }}"
       branch: "{{ ANALYTICS_SCHEDULE_SECURE_REPO_VERSION | default('master') }}"
       credential_id: "{{ ANALYTICS_SCHEDULE_SECURE_REPO_CREDENTIAL_ID | default('') }}"
+  analytics_tasks:
+    - id: ANSWER_DISTRIBUTION
+      name: AnswerDistributionWorkflow
```
Comment (mulby) [inline playbooks/roles/jenkins_analytics/defaults/main.yml]: I would prefer we leave the name definition in the DSL repo. I changed them, and likely wouldn't have thought to look here as well!

File: playbooks/roles/jenkins_analytics/tasks/main.yml [hunk 2]
```diff
@@ -95,43 +95,72 @@
...
+    mode: "600"
+    owner: "{{ jenkins_user }}"
+    group: "{{ jenkins_group }}"
+  tags:
+    - jenkins-seed-job
+
+- name: generate seed job xml
+  shell: >
+    cd {{ jenkins_seed_job_root }} &&
+    GRADLE_OPTS="-Dorg.gradle.daemon=true" ./gradlew run -Pargs={{ jenkins_seed_job_script }}
```
Comment (mulby) [inline playbooks/roles/jenkins_analytics/tasks/main.yml:180]: Is this only slow the first time and fast on subsequent builds? Or is it always slow?

Also, how long is **forever**?

**Comment by mulby** [comment]
@pomegranited I agree the seed job is now a lot easier to read / grok.

That said if this adds a huge amount of time to the build it's definitely something we should think about a bit.

@jibsheet what do you think?

**Comment by mulby** [comment]
Also it looks like our changes are somewhat overlapping: https://github.com/edx/configuration/pull/2962

File: playbooks/roles/jenkins_analytics/tasks/main.yml (hunk 2 above)
Comment (pomegranited) [inline playbooks/roles/jenkins_analytics/tasks/main.yml:180]: @mulby It took my little VM 7 minutes to do the initial gradle build to parse the seed job groovy into xml.  Subsequent builds are fast though, just a few seconds.

It's mainly frustrating that I can't figure out how to use the existing job-dsl-plugin .jar files to do this!

**Comment by pomegranited** [comment]
@mulby I won't try to merge this until after your #2962 goes in, so will apply your changes to it.

File: playbooks/roles/jenkins_analytics/defaults/main.yml (hunk 1 above)
Comment (pomegranited) [inline playbooks/roles/jenkins_analytics/defaults/main.yml]: @mulby I can generalise the [seed job parameter descriptions](https://github.com/edx/configuration/pull/2963/files#diff-5fc785f00323bc101492698cff6bb74dR41).. that's the only place this "name" field is used.

**Comment by jibsheet** [comment]
7 minutes on vagrant isn't _that_ bad, although it will slow initial setup for new developers and new processes.  If it produces a much more readable workflow, I think I'm ok with it being slower.

File: playbooks/roles/jenkins_analytics/tasks/main.yml (hunk 2 above)
Comment (jbzdak) [inline playbooks/roles/jenkins_analytics/tasks/main.yml:180]: For me the timings are 3 minutes on first run 30sec on every subsequent. While this is a long time, on initial run it doesn't add much to the overall time, and re-run that takes totally less than three minutes is acceptable.

File: playbooks/roles/jenkins_analytics/templates/seedJob.groovy [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  Jenkins Analytics Seed Job DSL template
+ */
+
+job('{{ jenkins_seed_job.name }}') {
+
+  description('Creates and configures the analytics task jobs.')
+
+  multiscm {
+  {%- for scm in jenkins_seed_job.multiscm -%}
+    {%- if scm.url -%}
```
Comment (jbzdak) [inline playbooks/roles/jenkins_analytics/templates/seedJob.groovy]: Are these `{%-` strictly needed, or is it for aesthetics?

Comment (pomegranited) [inline playbooks/roles/jenkins_analytics/templates/seedJob.groovy]: @jbzdak They were just for aesthetics, so I removed them with [a4cca36](https://github.com/edx/configuration/pull/2963/commits/a4cca360a8aca77bb22d8aa238b1a2223f55af1a).

File: playbooks/roles/jenkins_analytics/defaults/main.yml (hunk 1 above)
Comment (pomegranited) [inline playbooks/roles/jenkins_analytics/defaults/main.yml]: Removed the `name` field with [48c95cd](https://github.com/edx/configuration/pull/2963/commits/48c95cdcee4c2ee920341eb2682a6bed96d334fc)

**Comment by pomegranited** [comment]
@mulby I've merged your changes from #2962 into [a4cca36](https://github.com/edx/configuration/pull/2963/commits/a4cca360a8aca77bb22d8aa238b1a2223f55af1a) and rebased.

Let me know when it's ok to squash and merge?

**Comment by mulby** [comment]
:+1: 

If this runs cleanly on our install, I'm good with merging it. @jibsheet can you test it out?

**Comment by jibsheet** [comment]
@pomegranited I think you're good to merge 👍

**Pull request merged**